sdk/mobile-center-rum/src/main/java/com/microsoft/azure/mobile/rum/RealUserMeasurements.java

```diff
@@ -0,0 +1,475 @@
+package com.microsoft.azure.mobile.rum;
+
+import android.annotation.SuppressLint;
+import android.content.Context;
+import android.support.annotation.NonNull;
+import android.util.Log;
+
+import com.microsoft.azure.mobile.AbstractMobileCenterService;
+import com.microsoft.azure.mobile.MobileCenter;
+import com.microsoft.azure.mobile.channel.Channel;
+import com.microsoft.azure.mobile.http.DefaultHttpClient;
+import com.microsoft.azure.mobile.http.HttpClientNetworkStateHandler;
+import com.microsoft.azure.mobile.http.ServiceCallback;
+import com.microsoft.azure.mobile.utils.MobileCenterLog;
+import com.microsoft.azure.mobile.utils.NetworkStateHelper;
+import com.microsoft.azure.mobile.utils.UUIDUtils;
+import com.microsoft.azure.mobile.utils.async.MobileCenterFuture;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Random;
+
+import static com.microsoft.azure.mobile.http.DefaultHttpClient.METHOD_GET;
+
+/**
+ * RealUserMeasurements service.
+ */
+public class RealUserMeasurements extends AbstractMobileCenterService {
+
+    /**
+     * Name of the service.
+     */
+    private static final String SERVICE_NAME = "RealUserMeasurements";
+
+    /**
+     * TAG used in logging for Analytics.
+     */
+    private static final String LOG_TAG = MobileCenterLog.LOG_TAG + SERVICE_NAME;
+
+    /**
+     * Rum key length.
+     */
+    private static final int RUM_KEY_LENGTH = 32;
+
+    /**
+     * Rum configuration endpoint.
+     */
+    private static final String CONFIGURATION_ENDPOINT = "https://rumconfig.trafficmanager.net";
+
+    /**
+     * JSON configuration file name.
+     */
+    private static final String CONFIGURATION_FILE_NAME = "rumConfig.js";
+
+    /**
+     * Warm up image path.
+     */
+    private static final String WARM_UP_IMAGE = "trans.gif";
+
+    /**
+     * TestUrl image path.
+     */
+    private static final String SEVENTEENK_IMAGE = "17k.gif";
+
+    /**
+     * Flag to support https.
+     */
+    private static final int FLAG_HTTPS = 1;
+
+    /**
+     * Flag to use the 17k image to test.
+     */
+    private static final int FLAG_SEVENTEENK = 12;
+
+    /**
+     * Test url format.
+     */
+    private static final String TEST_URL_FORMAT = "http%s://%s/apc/%s?%s";
+
+    /**
+     * Report url format.
+     */
+    private static final String REPORT_URL_FORMAT = "https://%s?MonitorID=atm&rid=%s&w3c=false&prot=https&v=2017061301&tag=%s&DATA=%s";
+
+    /**
+     * Additional headers.
+     */
+    private static final Map<String, String> HEADERS = Collections.emptyMap();
+
+    /**
+     * Shared instance.
+     */
+    @SuppressLint("StaticFieldLeak")
```

Comment:
Maybe store context as `WeakReference<Context>` instead?

Reply:
Doc claims storing app context has issue with instant run but never saw any issue in reality. Passing everywhere to a weak reference could lead to null references if it ever invalidates to another context and we would need to manage that. I would prefer to have proofs that the app context gets updated for real before doing anything.